Monika-After-Story/MonikaModDev
# Dynamic Empty Desk

#3820

# NOTE:
**I have added TODOs for weather-related changes, these will be addressed in a different PR** (link to be added when that PR is made)

# Key Features
* Adds new empty desk sprites from @Velius94 
* Empty desk changed from a basic sprite to a dynamic (+ acs) sprite that includes shadow and the chair as separate configrables
* Zorder of the background moved to 3 in preparation for room decoration
* new ACS property + json `keep_on_desk` which marks that an ACS should remain on desk if Monika is not there.  (all mug/choco/plush/desk acs have this set to True)
* new `ACSTemplate` property `keep_on_desk` which can be used to default `keep_on_desk` for certain ACS by type. (currently this is used for mugs)
* **empty desk transitions no longer require zoom to be reset**
* added parameter `show_emptydesk` to `spaceroom` label. it's behavior is dependent on the `hide_monika` parameter (see the label for more info). this label also (by default) will show/hide emptydesk if monika is hidden/shown.

## Empty Desk
Empty desk is no longer a ConditionSwitch image. It is now a DynamicDisplayable with the `mas_drawemptydesk` function. This function pulls from `MASMonika`. 

Additions:
* `store.mas_sprites.show_empty_desk` - function that shows the emptydesk wtih appropriate zorders and stuff (via renpy.show)
* `mas_transition_to_emptydesk` - label that shows the emptydesk and hides monika with dissolve
* `mas_transition_from_emptydesk` - label that hides emptydesk and shows monika with dissolve. Accepts an expression string parameter that can be used to show monika with a specific expression

### Sprites
Chair sprites are always rendered after back-hair, before body-0. They should be of the format `chair-<tag><night suffix>`

Table sprites will render with a shadow only if monika is present and the correct shadow sprites are available. The table should be of the format `table-<tag><night suffix>`. The shadow should be `table-<tag>-s<night suffix>`.

### MASMonika
* new property `tablechair` - contains a `MASTableChair` object which is used for configuring a table/chair combination
* `get_acs_by_desk` - function that retrieves a list of ACS objects based on their `keep_on_desk` flag

### MASTableChair
* representation of a "Desk" (table and chair). **there should only be one of these objects**. If we want to change the table/chair, modify `monika_chr.tablechair`.
* always use `setTable` when setting the table tag. This function calls `self.prepare()` which checks if the table has shadow sprites. this prevents potential crashes from table sprites not having shadows.
* `table` - property that contains the tag used in building the table sprite string
* `chair` - property that contains the tag used in building the chair sprite string
* `has_shadow` - True if this table has a shadow sprite, False if not. **DO NOT CHANGE** this is set by the `prepare` and `setTable` functions.

### MASAccessory
* new `keep_on_desk` property that determines if a an ACS should remain if Monika leaves. This will be set by default for ACS that should stay on the desk. However, to make an ACS leave if Monika leaves (like getting another cup of coffee), set the `keep_on_desk` flag to False before showing emptydesk

# Testing
Use dev topics and verify appearance of empty desk makes sense. (Shadow should go away when Monika goes away, empty desk should respect zoom level with a minor descrepancy)
* `TEST EMPTY DESK` - tests empty desk visuals
* `TEST EMPTY DESK (with transition labels)` - tests empty desk visuals using the transition labels

The following locations have been affected by empty desk and should be tested:
* final farewell - the empty desk should not have any ACS on it and should be 
* all o31 costume greetings - ignore marisa and rin for now.
* labels `mas_monika_gets_cake` and `mas_hideCake` - not sure how these are suppoesd to be reached so onto you guys for figuring out workflow
* `The Hunter` scary story 
* coffee/hot chocolate drinking - verify the first-time gift reaction, the standard spaceroom idle ones, and the idle mode ones. in all these cases, we should **not** be zooming out anymore
* anywhere we change clothes via `mas_clothes_change` - again, zoom is no longer an thing
* empty desk when taking monika out but not returning her before opening game

Other things:
* verify the opendoor greeting still looks ok - the joke with the window masks should still work.
* verify that poses look okay. There was some refactoring of sprite string generation
* verify all locations that call spaceroom are still okay.
* verify lightning still looks okay
